blog/2021/08/DEQ.md

```diff
@@ -0,0 +1,426 @@
+much of the ideas from the NeuralODE line of research, and allows neural network structures
+to be viewed as repeated applications of layered computations. This is often very accurate,
+for example when we apply convolution filters on images---the network can essentially
+be viewed as a convolutional feature extractor consisting of repetitive blocks, and one
+linear output layer at the very end. When there is an infinite number of these repeated
+blocks, the network might explode off to infinity. If it doesn't, however, it reaches an
+equilibrium points and converges, hence the name.
+
+## DEQ Models and Infinitely Deep Networks
+
+Let us begin by looking at an instantiation of the equations for DEQ models.
+$$
+\begin{array}{l}
+z:=0\\
+\text { Repeat until convergence: }\\
+z:=\tanh (W z+x)
+\end{array}
+$$
+
+Which eventually brings us to
+$$
+z^{\star}=\tanh \left(W z^{\star}+x\right)
+$$
+
+The idea of DEQ is, in a sense, insanely simple! We assume there are infinitely many
+compositions of identical layers (equivalently an infinitely deep network), and
+directly solve for the fixed point described by this equation. Clearly, naively iterating
+for the fixed point would take an enormous amount of time. Thanks to the Implicit Function
+Theorem, DEQ shows that it is possible to train an infinitely-deep neural network using the
+gradient optimization toolbox. For more reference, please see
+[Deep Implicit Layers](http://implicit-layers-tutorial.org)
+
+From the viewpoint of Julia and the DiffEqFlux.jl library, it is also natural to look at DEQ from
+a differential equations perspective. Thanks to NeuralODE, the machine learning community is now able
+to put on the glasses of differential equations, and see through a new perspective.
+If we perceive the iterating layer of the network as an ODE, DEQ is solving nothing but the
+well-known "Steady State Problem", defined as the final converging state of the ODE as time
+goes to infinity. This line of work is well-studied in the scientific machine learning
+literature, see for example [Notes on Adjoint Methods for 18.335](https://math.mit.edu/~stevenj/18.336/adjoint.pdf)
+
+$$
+f(u) - u = du = 0
+$$
+
+Reaching a fixed point of the iterative layer application is much like solving for the final
+value in a converging sequence. To put it simply, we set the current value equal to the next,
+and obtain a solution for this recursion problem. As we show above, if we use `u` to represent
+the current layer and `f(u)` to represent the next layer, we can sloppily denote the fixed point
+equation as `f(u)-u=0`. Crucially, this can also be converted into a steady state problem
+by defining an ODE layer `f(u)-u=du` and solving
+that ODE in its steady state. This gives us equality among all three components.
+
+## Let us implement a simple DEQ Model
+
+The following code block creates a DEQ model. The acute will notice that this code looks
```

Comment:
```suggestion
There are many ways that one can solve a rootfinding problem with different characteristics.
One can directly use Newton's method, but this can require a good guess and may not distinguish
between stable and unstable equilibrium. Bifurcation tools like [BifurcationKit.jl](https://rveltz.github.io/BifurcationKit.jl/dev/) give a whole host of other methods. But for our demonstration,
we will focus on showcasing the evolving ODE approach. This is similar to the fixed point method of running
$x_{n+1} = x_n + f(x_n)$, but we can instead allow the ODE solver to adaptively change its steps along the
trajectory to better facilitate convergence. We can define "convergence" as a tolerance where for $x' = f(x)$,
we want $x'$ to be sufficiently small (in absolute and relative tolerance). It turns out that a method for
doing this, along with its backpropogation, is already defined in Julia as the 
[steady state problem](https://diffeq.sciml.ai/stable/types/steady_state_types/) which will automatically
detect when the ODE has converged and halt the integration. So let's demonstrate a DEQ via SteadyStateProblem.

The following code block creates a DEQ model. The acute will notice that this code looks
```